### DISCUSSION TITLE
The userId set through TraceContext.putCorrelation("user", userId.toString()) in the gateway is not propagated to the downstream system.

### DISCUSSION THREAD

**Discussion by hgaong** [apache/skywalking#11367] [Q&A] [answered]
<img width="610" alt="image" src="https://github.com/apache/skywalking/assets/29444991/a087428b-e651-4339-bf79-014f580ca6b4">
<img width="498" alt="image" src="https://github.com/apache/skywalking/assets/29444991/aca9230e-3bf1-480e-8cce-6cc2107c9bde">
<img width="602" alt="image" src="https://github.com/apache/skywalking/assets/29444991/e22f16ca-cc9f-4473-b8d9-fce6b7c53349">

In agent version 9.0, when trying to retrieve the correlation value using TraceContext.getCorrelation("user") in the downstream system, it returns an empty result.

agent log error :
<img width="1378" alt="image" src="https://github.com/apache/skywalking/assets/29444991/100e7a3d-4c8d-42e8-98f4-80cc483fe58f">
gateway log :
<img width="1347" alt="image" src="https://github.com/apache/skywalking/assets/29444991/80c93a7f-de5b-41cd-a4bc-a82d133cbf13">

gateway code :
<img width="774" alt="image" src="https://github.com/apache/skywalking/assets/29444991/d76af196-8700-4dc2-ab08-6b45487b1d16">

**Comment by wu-sheng** [comment]
Does your tracing continue? If the trace breaks, this breaks.

**Reply by hgaong** [reply]
<img width="532" alt="image" src="https://github.com/apache/skywalking/assets/29444991/0e4f71cb-4ea8-487c-965a-98a7a1bf0fe5">
<img width="1301" alt="image" src="https://github.com/apache/skywalking/assets/29444991/57ab7e20-8f70-4ce5-90e3-9112d5217ca9">


After adding the @trace annotation to my tagUserId method, there was no NullPointerException in the code line within agent.

If the @Trace annotation is not added, there would be a NullPointerException in the code line within agent:
![image](https://github.com/apache/skywalking/assets/29444991/eda6c899-dca4-4c79-ab39-7f94b9087abe)


The tagUserId method is called within a custom filter in Spring Cloud Gateway. Is it due to incorrect calling position? Does the ContextManager.getCorrelationContext() return null when called? Where should it be called to avoid returning null?

**Reply by wu-sheng** [reply]
Spring Gateway is complex as always. I don't know enough about that, as it is just a simple plugin from a maintainer's perspective. However the end users of Gateway have various ways of using it. So, from what I saw, your way is not supported.

You need to debug the plugins and understand the codes by yourself. Or, just wait for someone from the community have known what this means. But most likely, this would not be from a maintainer. AFAIK, no maintainer is working on any specific plugin.

**Comment by hgaong** [comment]
Thank you very much for taking the time to answer my questions. Most of the content available online about Skywalking is focused on basic setup tutorials, and the few articles on advanced features are vague and lack practical reference value. The official documentation for the Java agent is also quite simple. It seems that I'll have to debug it myself. In any case, I'm really grateful for your help.




---Original---
From: "吴晟 Wu ***@***.***&gt;
Date: Tue, Oct 3, 2023 08:34 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [apache/skywalking] The userId set throughTraceContext.putCorrelation("user", userId.toString()) in the gateway is notpropagated to the downstream system. (Discussion #11367)




 
Spring Gateway is complex as always. I don't know enough about that, as it is just a simple plugin from a maintainer's perspective. However the end users of Gateway have various ways of using it. So, from what I saw, your way is not supported.
 
You need to debug the plugins and understand the codes by yourself. Or, just wait for someone from the community have known what this means. But most likely, this would not be from a maintainer. AFAIK, no maintainer is working on any specific plugin.

**Reply by wu-sheng** [reply]
The docs for users are simple and straight forward. That is on purpose.

Once you need to dive in, dev docs and source codes are keys for open sources.

Such as for Java agent, https://skywalking.apache.org/docs/skywalking-java/next/en/setup/service-agent/java-agent/java-plugin-development-guide/

**Comment by weixiang1862** [comment, marked as answer]
@hgaong `TraceContext.putCorrelation("xx", "xx")` need an active span in current thread. In custom gateway filter, it does not exist.

spring-cloud-gateway use webflux, it propagate tracingcontext by dynamicfield on ServerWebExchange, you can change your code like this:
```
@Component
public class AuthFilter implements GlobalFilter, Ordered {

    @Override
    public Mono<Void> filter(final ServerWebExchange exchange, final GatewayFilterChain chain) {
   ---> WebFluxSkyWalkingTraceContext.putCorrelation(exchange, "user","xxx");
        return chain.filter(exchange);
    }

    @Override
    public int getOrder() {
        return 0;
    }
}
```
WebFluxSkyWalkingTraceContext provided by `apm-toolkit-webflux`, do not forget add it to pom.xml:
```
<dependency>
    <groupId>org.apache.skywalking</groupId>
    <artifactId>apm-toolkit-webflux</artifactId>
    <version>${version}</version>
</dependency>
```
for more details, you can ref: [WebFluxSkyWalkingCorrelationContextPutInterceptor.java](https://github.com/apache/skywalking-java/blob/main/apm-sniffer/apm-toolkit-activation/apm-toolkit-webflux-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/webflux/WebFluxSkyWalkingCorrelationContextPutInterceptor.java)

And there is an chinese article about correlation write by someone, I think it can help you understand it well: 
[【SkyWalking】巧用SkyWalking实现全链路传递userId等自定义数据](
https://juejin.cn/post/7282687884205129787?searchId=20231009171236457BA379F504A618AF0)
